nlmod/cache.py

```diff
@@ -915,3 +935,115 @@ def _explicit_dataset_coordinate_comparison(ds_in, ds_cache):
             return False
     logger.debug("cache -> all coordinates equal")
     return True
+
+
+class NumpyEncoder(json.JSONEncoder):
+    """Special json encoder for numpy types."""
+
+    def default(self, obj):
+        if isinstance(obj, np.integer):
+            return int(obj)
+        elif isinstance(obj, np.floating):
+            return float(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def hash_xarray_coords(ds, include_metadata: bool = False):
+    """
+    Create a hash of xarray coordinate(s) using array bytes and optionally metadata.
+
+    Parameters
+    ----------
+    coord : xarray.core.coordinates.Coordinate
+        The xarray coordinate object to hash
+    include_metadata : bool, optional
+        Whether to include metadata in the hash. Default is False.
+
+    Returns
+    -------
+    str
+        The hexadecimal hash string
+    """
+    combined_bytes = b""
+    for coord in ds.coords:
+        # get the raw bytes from the numpy array values
+        values_bytes = ds[coord].values.tobytes()
+        combined_bytes += values_bytes
+
+        if include_metadata:
+            # get metadata as JSON
+            metadata = {
+                "name": coord,
+                "dims": ds[coord].dims,
+                "attrs": ds[coord].attrs,
+                "dtype": str(ds[coord].dtype),
+                "shape": ds[coord].shape,
+            }
+            metadata_bytes = json.dumps(
+                metadata, sort_keys=True, cls=NumpyEncoder
+            ).encode("utf-8")
+
+            # combine both sets of bytes for hashing
+            if include_metadata:
+                combined_bytes += metadata_bytes
+
+    # create a hash of the combined bytes
+    return hashlib.sha256(combined_bytes).hexdigest()
+
+
+def hash_xarray_data_vars(
+    ds,
+    include_metadata: bool = False,
+):
+    """
+    Create a hash of xarray data variables using array bytes and optionally metadata.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset or xarray.DataArray
+        The xarray coordinate object to hash
+    include_metadata : bool, optional
+        Whether to include metadata in the hash. Default is False.
+
+    Returns
+    -------
+    str
+        The hexadecimal hash string
+    """
+    # get the raw bytes from the numpy array values
+    combined_bytes = b""
+    if isinstance(ds, xr.Dataset):
+        data_arrays = [ds[da] for da in ds.data_vars]
```

Comment:
Do we have to worry about the ds being sorted?

Reply:
Good question, it would be prudent to sort to make sure, as you said.